### ISSUE TITLE
using host netns is insecure

### ISSUE THREAD

**Issue by AkihiroSuda** [cyphar/rootlesscontaine.rs#5]
@cyphar 

https://www.cyphar.com/blog/post/20160627-rootless-containers-with-runc

> There’s no security issue with not using a namespace, since we never have any privileges when setting up a rootless container.

It has issues.
For example, abstract UNIX sockets are leaked to containers.

```
terminal1-host$ socat ABSTRACT-LISTEN:foobar STDIO
```

```
terminal2-host$ runc run some-rootless-container
terminal2-runc# echo hacked | socat STDIO ABSTRACT-CONNECT:foobar
```

### Use suid helper binary

This should be the easiest and the best in performance.
Actually, LXC has been already doing this for years.

However, I don't think we prefer suid binaries.

### Ptrace

This should be possible but the worst, because it is hard to implement, in all aspects of  security, performance, and stability.

We could use seccomp for acceleration, but I could not get it work https://github.com/AkihiroSuda/runrootless/issues/4

Also, I don't think this works well with future CNI integration.

### Netfilter queue

I confirmed `brctl addbr foo` and `iptables -A OUTPUT -j NFQUEUE --queue-num 42`  are allowed in unprivileged netns. 
(Tested Ubuntu 17.10 and openSUSE Tumbleweed)

So we could inject helper processes to inside and outside the netns, and connect them via some IPC.
The helper process outside the netns parses, modifies, and routes L3 packets.

This would be slow, but there should be good room for optimization.

I'll try to implement POC of this when I can make time.

**Comment by AkihiroSuda** [comment]
Also, I think we should be able to tun/tap. (still trying)

**Comment by cyphar** [comment]
> It has issues. For example, abstract UNIX sockets are leaked to containers.

I don't agree this is a security issue. Yes, the socket is visible in the container, but it is also true that `127.0.0.1` bindings will also be visible from inside the container when you don't unshare the network namespace. While you have access to them, you haven't gained any privileges as a result. It's definitely not ideal, but the key phrase I used was "*security* issue", and I don't think this qualifies.

However, I do agree that work needs to happen to fix this issue.

> So we could inject helper processes to inside and outside the netns, and connect them via some IPC.
The helper process outside the netns parses, modifies, and routes L3 packets.

This was an idea I pitched a while ago, the one thing that annoyed me was that you couldn't match a `bind(2)` call in the container to one on the host (because it doesn't go through NFQUEUE) but I guess that's not a blocking issue. Though from memory there might've been some blocking (permission-related) issue that stopped this from being viable -- I might be mistaken though.

However, if this actually can work, I'd be more than happy to work with you on it. We could make it compatible with OCI hooks, so we can have unprivileged network setups work out-of-the-box.
